### ISSUE TITLE
Predicate pushdown can prevent redundant join from doing its job

### ISSUE THREAD

**Issue by wangandi** [MaterializeInc/materialize#6748]
`create table t1(f1 int, f2 int, f3 int);`

In the query below, our `RedundantJoin` optimization successfully removes the join:
```
materialize=> explain 
select * 
from t1 a1, (select f1, f2 from t1 group by f1, f2) a2 
where (a1.f1 = a2.f1 or (a1.f1 is null and a2.f1 is null))
  and (a1.f2 = a2.f2 or (a1.f2 is null and a2.f2 is null));
          Optimized Plan           
-----------------------------------
 %0 =                             +
 | Get materialize.public.t1 (u53)+
 | Project (#0..#2, #0, #1)       +
```

However, if you tweak the query slightly, `RedundantJoin` becomes unable to remove the join because a filter has been pushed to each input of the join. 
```
materialize=> explain 
select * 
from t1 a1, (select f1, f2 from t1 group by f1, f2) a2 
where (a1.f1 = a2.f1 or (a1.f1 is null and a2.f1 is null))
  and a1.f2 = a2.f2;
                  Optimized Plan                  
--------------------------------------------------
 %0 =                                            +
 | Get materialize.public.t1 (u53)               +
 | Filter !(isnull(#1))                          +
                                                 +
 %1 =                                            +
 | Get materialize.public.t1 (u53)               +
 | Filter !(isnull(#1))                          +
 | Distinct group=(#0, #1)                       +
 | ArrangeBy (#0, #1)                            +
                                                 +
 %2 =                                            +
 | Join %0 %1 (= #0 #3) (= #1 #4)                +
 | | implementation = Differential %0 %1.(#0, #1)+
 | | demand = (#0..#2)                           +
 | Project (#0..#2, #0, #1)                      +
```

This issue cannot be solved by running `RedundantJoin` before `PredicatePushdown`. `RedundantJoin` cannot remove the join until it knows the equivalence classes of the join. `PredicatePushdown` sets the equivalence classes of the join at the same time that it pushes the filter down to both inputs of the join.

**Comment by wangandi** [comment]
One fix would be to enable `RedundantJoin` to use subquery matching (#716).

**Comment by wangandi** [comment]
It appears that running `InlineLet` too often/frequently also prevents redundant join from doing its job. RedundantJoin is capable of removing redundant inputs with filters if they are wrapped up in a let. For example, RedundantJoin would work if a `MirRelationExpr` looks like:
```
 %0 = Let l0 =                                        +
 | Get materialize.public.t1 (u53)               +
 | Filter !(isnull(#1))                          +
                                                 +
 %1 =                                            +
 | Get l0               +
 | Distinct group=(#0, #1)                       +
 | ArrangeBy (#0, #1)                            +
                                                 +
 %2 =                                            +
 | Join %0 %1 (= #0 #3) (= #1 #4)                +
```

**Comment by asenac** [comment]
I ran into this problem in #7087. One possible solution would be to add a new transform that is always executed before redundant join that would do the opposite of predicate pushdown. I would lift all predicates are far up as possible, ie. without crossing any error boundary. Then, after running redundant join transform, we would run predicate pushdown again, which should leave all predicates where they were.

**Comment by frankmcsherry** [comment]
To reasonable fixes here:

1. allow `RedundantJoin` to consider more things than `Get` statements. 
2. Implement `MirRelationExpr` CSE, which would capture equivalent subexpressions and bundle them behind a `Get`.

**Comment by frankmcsherry** [comment]
This is fixed in #7715:
```
materialize=> explain 
select * 
from t1 a1, (select f1, f2 from t1 group by f1, f2) a2 
where (a1.f1 = a2.f1 or (a1.f1 is null and a2.f1 is null))
  and a1.f2 = a2.f2;
          Optimized Plan          
----------------------------------
 %0 =                            +
 | Get materialize.public.t1 (u5)+
 | Filter !(isnull(#1))          +
 | Project (#0, #1, #0, #1)      +
 
(1 row)

materialize=> 
```
